src/sign_tx.c

```diff
@@ -0,0 +1,506 @@
+        // 0x01 is the left icon so return NULL if we're displaying the beginning of the text.
+        return (ctx->display_index == 0) ? NULL : element;
+    case 2:
+        // 0x02 is the right, so return NULL if we're displaying the end of the text.
+        return ctx->display_index == (strlen((const char*)ctx->info) - MAX_SCREEN_LENGTH) ? NULL : element;
+    default:
+        // Always display all other elements.
+        return element;
+    }
+}
+
+// This is the button handler for the outputs screen.
+static unsigned int ui_sign_tx_compare_button(unsigned int button_mask, unsigned int button_mask_counter) {
+    switch (button_mask) {
+        // scroll left by either clicking or pressing the left button
+        case BUTTON_LEFT:
+        case BUTTON_EVT_FAST | BUTTON_LEFT: // SEEK LEFT
+            if (ctx->display_index != 0) {
+                ctx->display_index--;
+                os_memmove(ctx->line2, ctx->info + ctx->display_index, MAX_SCREEN_LENGTH);
+                UX_REDISPLAY();
+            }
+
+            break;
+
+        // scroll right by either clicking or pressing the right button
+        case BUTTON_RIGHT:
+        case BUTTON_EVT_FAST | BUTTON_RIGHT: // SEEK RIGHT
+            if (ctx->display_index != strlen((const char*)ctx->info) - MAX_SCREEN_LENGTH) {
+                ctx->display_index++;
+                os_memmove(ctx->line2, ctx->info + ctx->display_index, MAX_SCREEN_LENGTH);
+                UX_REDISPLAY();
+            }
+
+            break;
+
+        case BUTTON_EVT_RELEASED | BUTTON_LEFT | BUTTON_RIGHT: // PROCEED TO NEXT OUTPUT
+            ctx->display_index = 0;
+            switch(decode_next_element()) {
+                case TX_STATE_ERR:
+                    io_exchange_with_code(SW_INVALID_PARAM, 0);
+                    ui_idle();
+                    break;
+                case TX_STATE_PARTIAL:
+                    // We don't have enough data to decode the next element; send an
+                    // OK code to request more.
+                    io_exchange_with_code(SW_OK, 0);
+                    break;
+                case TX_STATE_READY:
+                    //display element
+                    if (ctx->elem_type == ELEM_OUTPUT) {
+                        prepare_display_output(ctx->decoded_output);
+                    } else {
+                        THROW(SW_INVALID_PARAM);
+                    }
+
+                    UX_REDISPLAY();
+                    break;
+                case TX_STATE_FINISHED:
+                    // Go to confirmation screen
+                    os_memmove(ctx->line1, "Send\0", 5);
+                    os_memmove(ctx->line2, "transaction?\0", 13);
+                    UX_DISPLAY(ui_sign_tx_confirm, ui_prepro_sign_tx_confirm);
+                    break;
+            }
+    }
+    return 0;
+}
+
+void handle_sign_tx(uint8_t p1, uint8_t p2, uint8_t *data_buffer, uint16_t data_length, volatile unsigned int *flags, volatile unsigned int *tx) {
```

Comment:
It seems you can split this method into multiple internal methods and just call them according to each case. It would be easier to read the method and also to test if we need to.

Reply:
Done